piksi_tools/fileio.py

```diff
@@ -10,209 +11,416 @@
+        self._request_pool = Queue(window_size)
+        for pending_req in self._pending_map:
+            self._request_pool.put(pending_req)
+        self._batch_size = batch_size
+
+    def __enter__(self):
+        self._link.add_callback(self._request_cb, self._msg_type)
+        self._link.add_callback(self._config_cb, SBP_MSG_FILEIO_CONFIG_RESP)
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self._link.remove_callback(self._request_cb, self._msg_type)
+        self._link.remove_callback(self._config_cb, SBP_MSG_FILEIO_CONFIG_RESP)
 
     def _verify_cb_thread(self):
         """
-        Verify that the same (singular) thread is accessing the `self.pending`
-        and the `self.write_pool` lists.  Only the cb thread should free window
-        space by removing (popping) from `self.pending` and appending it to the
-        `self.write_pool` list.
+        Verify that only one thread is consuming requests.
         """
-        if self.cb_thread is None:
-            self.cb_thread = threading.currentThread().ident
-        assert self.cb_thread == threading.currentThread().ident
+        if self._callback_thread is None:
+            self._callback_thread = threading.currentThread().ident
+        assert self._callback_thread == threading.currentThread().ident
 
     def _verify_link_thread(self):
         """
-        Verify that the same (singular) thread is accessing the `self.pending`
-        and the `self.write_pool` lists.  Only the link thread should consume
-        window by removing a PendingWrite object from the `self.write_pool`
-        list and appending it to the `self.pending` list.
+        Verify that only one thread is producing requests.
         """
-        if self.link_thread is None:
-            self.link_thread = threading.currentThread().ident
-        assert self.link_thread == threading.currentThread().ident
+        if self._link_thread is None:
+            self._link_thread = threading.currentThread().ident
+        assert self._link_thread == threading.currentThread().ident
 
-    def _return_pending_write(self, pending_write):
+    def _return_pending_req(self, pending_req):
         """
-        Increases the count of available pending writes by move a PendingWrite
-        object from `self.pending` to `self.write_pool`.
-
-        Threading: only the callback thread should access this function.  The
-        cb thread is the consumer of `self.pending`, and the producer of
-        `self.write_pool`.
+        Return a pending request to the write pool and clean any
+        entries in the expiration map.
         """
         self._verify_cb_thread()
-        self.pending.pop(self.pending.index(pending_write))
-        self.write_pool.append(pending_write.invalidate())
+        self._request_pool.put(pending_req)
+        del self._seqmap[pending_req.message.sequence]
+        del self._expire_map[pending_req.time_expire][pending_req]
```

Comment:
when testing with vanilla 1.5/2.0/2.1/2.2 I'm getting quite a few of these:
```
  File "piksi_tools/fileio.py", line 262, in _return_pending_req
    del self._expire_map[pending_req.time_expire][pending_req]
  KeyError: PendingRequest(offset=18924416,seq=833422368,time=Time(s=1549918415,ms=12),tries=0,index=58)
```
although transfers seem to complete (although I'm testing with a junk file, verification is not being tested). maybe the `del` needs to be wrapped with some other logic?

Reply:
Actually I just got stuck on a transfer with 2.2 but it's really hard to tell if this was a pre-existing condition:
![screen shot 2019-02-11 at 1 41 02 pm](https://user-images.githubusercontent.com/5225320/52595480-d8958280-2e02-11e9-898b-f018abb1e44a.png)
May not be actionable for this PR but wanted to note since I took the time to check these out.

Reply:
Not sure how this could happen, unless we somehow lost a chunk of the file that we thought timed out and never requested a retry.

Reply:
I think I found the issue (I was able to reproduce the hang by running transfers in an infinite loop), there's a small change that we could "return" a pending request that was already returned (not exactly sure how)-- but fixed here 9a797c2b251772745fbe0bae0eeaedadf2fb28d1.  Ran a transfer in a loop overnight and didn't see the hang happen again.